superset/db_engine_specs/duckdb.py

```diff
@@ -0,0 +1,103 @@
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+import re
+from datetime import datetime
+from typing import Any, Dict, List, Optional, Pattern, Tuple, TYPE_CHECKING
+
+from flask_babel import gettext as __
+from sqlalchemy.engine.reflection import Inspector
+
+from superset.db_engine_specs.base import BaseEngineSpec
+from superset.errors import SupersetErrorType
+from superset.utils import core as utils
+
+if TYPE_CHECKING:
+    # prevent circular imports
+    from superset.models.core import Database
+
+
+COLUMN_DOES_NOT_EXIST_REGEX = re.compile("no such column: (?P<column_name>.+)")
+
+
+class DuckDBEngineSpec(BaseEngineSpec):
+    engine = "duckdb"
+    engine_name = "DuckDB"
+
+    # pylint: disable=line-too-long
+    _time_grain_expressions = {
+        None: "{col}",
+        "PT1S": "DATE_TRUNC('second', {col})",
+        "PT1M": "DATE_TRUNC('minute', {col})",
+        "PT1H": "DATE_TRUNC('hour', {col})",
+        "P1D": "DATE_TRUNC('day', {col})",
+        "P1W": "DATE_TRUNC('week', {col})",
+        "P1M": "DATE_TRUNC('month', {col})",
+        "P0.25Y": "DATE_TRUNC('quarter', {col})",
+        "P1Y": "DATE_TRUNC('year', {col})",
+    }
+
+    custom_errors: Dict[Pattern[str], Tuple[str, SupersetErrorType, Dict[str, Any]]] = {
+        COLUMN_DOES_NOT_EXIST_REGEX: (
+            __('We can\'t seem to resolve the column "%(column_name)s"'),
+            SupersetErrorType.COLUMN_DOES_NOT_EXIST_ERROR,
+            {},
+        ),
+    }
+
+    @classmethod
+    def epoch_to_dttm(cls) -> str:
+        return "datetime({col}, 'unixepoch')"
+
+    @classmethod
+    def get_all_datasource_names(
+        cls, database: "Database", datasource_type: str
+    ) -> List[utils.DatasourceName]:
+        schemas = database.get_all_schema_names(
+            cache=database.schema_cache_enabled,
+            cache_timeout=database.schema_cache_timeout,
+            force=True,
+        )
+        schema = schemas[0]
+        if datasource_type == "table":
+            return database.get_all_table_names_in_schema(
+                schema=schema,
+                force=True,
+                cache=database.table_cache_enabled,
+                cache_timeout=database.table_cache_timeout,
+            )
+        if datasource_type == "view":
+            return database.get_all_view_names_in_schema(
+                schema=schema,
+                force=True,
+                cache=database.table_cache_enabled,
+                cache_timeout=database.table_cache_timeout,
+            )
+        raise Exception(f"Unsupported datasource_type: {datasource_type}")
```

Comment:
This part looks like it can directly inherit the [Super Class](https://github.com/apache/superset/blob/e1d0b83885d89c50e96c6eef872beffe0de7cf50/superset/db_engine_specs/base.py#L825).

Reply:
Hmm, good catch - 
https://github.com/apache/superset/blob/master/superset/db_engine_specs/base.py#L825
incidentally, looks like the SQLite datasource follows the same pattern
https://github.com/apache/superset/blob/master/superset/db_engine_specs/sqlite.py#L74
I don't think it would be wise to clean that up in this PR but might be a quick win.

Reply:
made the change here - https://github.com/rwhaling/superset/commit/9fa02d7c1cc4b1c672f5f68637b3c922bfae32d6
I forgot that I can't point an existing PR to a new branch, might need to close this one - will wait for any additional feedback first though.  Still working on getting set up for local testing as well.